File: mods/faculator-expoter/exporter/technology.lua
--- 科技导出
--- 导出所有 LuaTechnologyPrototype，包含前置科技、解锁效果、研究材料等。

local log  = require("lib.logger").create("科技导出")
local Util = require("lib.util")

local Technology = {}

--- 导出所有科技原型
--- @see https://lua-api.factorio.com/latest/classes/LuaTechnologyPrototype.html
--- @return table[] 按名称排序的科技数组
function Technology.export()
  local safe = Util.safe_get
  local result = {}

  for name, proto in pairs(prototypes.technology) do
    -- 科技效果（解锁配方、修改器等）
    local effects = {}
    if proto.effects then
      for _, eff in ipairs(proto.effects) do
        local effect = { type = eff.type }
        if eff.type == "unlock-recipe" then
          effect.recipe = eff.recipe
        elseif eff.type == "unlock-space-location" then
          effect.space_location = safe(function() return eff.space_location end)
        else
          effect.modifier = safe(function() return eff.modifier end)
        end
        effect.quality = safe(function() return eff.quality end)
        effect.change = safe(function() return eff.change end)
        effects[#effects + 1] = effect
      end
    end

    -- 前置科技
    --- @see https://lua-api.factorio.com/latest/classes/LuaTechnologyPrototype.html#prerequisites
    local prerequisites = {}
    if proto.prerequisites then
      for _, prereq in pairs(proto.prerequisites) do
        prerequisites[#prerequisites + 1] = prereq.name
      end
      table.sort(prerequisites)
    end

    -- 研究材料
    --- @see https://lua-api.factorio.com/latest/classes/LuaTechnologyPrototype.html#research_unit_ingredients
    local ingredients = {}
    if proto.research_unit_ingredients then
      for _, ing in ipairs(proto.research_unit_ingredients) do
        ingredients[#ingredients + 1] = {
          type   = ing.type,
          name   = ing.name,
          amount = ing.amount,
        }
      end
    end

    -- 研究触发器（部分科技由事件触发而非消耗科学包）
    local research_trigger = safe(function()
      local rt = proto.research_trigger
      if not rt then return nil end
      return {
        type   = rt.type,
        entity = safe(function()
          local entity = rt.entity
          return entity and entity.name or entity
        end),
        item = safe(function()
          local item = rt.item
          return item and item.name or item
        end),
        fluid = safe(function()
          local fluid = rt.fluid
          return fluid and fluid.name or fluid
        end),
        count  = safe(function() return rt.count end),
        amount = safe(function() return rt.amount end),
      }
    end)

    result[#result + 1] = {
      name     = name,
      group    = proto.group.name,
      subgroup = proto.subgroup.name,
      order    = proto.order,
      hidden   = proto.hidden or false,

      essential              = safe(function() return proto.essential end),
      enabled                = proto.enabled,
      visible_when_disabled  = proto.visible_when_disabled,
      upgrade                = proto.upgrade,
      level                  = proto.level,
      max_level              = proto.max_level,

      --- 研究消耗
      research_unit_count         = safe(function() return proto.research_unit_count end),
      research_unit_count_formula = safe(function() return proto.research_unit_count_formula end),
      research_unit_energy        = proto.research_unit_energy,
      research_unit_ingredients   = ingredients,
      research_trigger            = research_trigger,

      prerequisites = prerequisites,
      effects       = effects,

      --- 是否允许产能研究
      allows_productivity = safe(function() return proto.allows_productivity end),

      --- 后续科技
      successors = safe(function()
        local succ = proto.successors
        if not succ then return nil end
        local res = {}
        for _, s in pairs(succ) do
          res[#res + 1] = s.name
        end
        table.sort(res)
        return res
      end),
    }
  end
  table.sort(result, function(a, b) return a.name < b.name end)
  log.info("导出科技: " .. #result .. " 个")
  return result
end

return Technology

File: crates/library/data/src/game_data/technology.rs
//! 顶层 `technologies` DTO。
//!
//! 科技对象是导出里关系最密集的一类原型之一：
//!
//! - `prerequisites[]` 指向其他科技
//! - `successors[]` 反向指向后继科技
//! - `effects[]` 解锁 recipe / quality / space_location 或修改某种游戏能力
//! - `research_unit_ingredients[]` 指向科技包物品
//!
//! 此外，导出器在这里使用了一个典型“小巧思”：
//!
//! - 当某个数组为空时，导出有时会给出 `{}` 而不是 `[]`
//!
//! 本模块会通过显式 DTO 类型保留这种双形态，避免在反序列化层做黑盒规整。
use crate::game_data::ArrayOrEmptyObject;
use serde::{Deserialize, Serialize};

mod effect;
mod effect_modifier;
mod research_trigger;
mod research_unit_ingredient;

pub use effect::*;
pub use effect_modifier::*;
pub use research_trigger::*;
pub use research_unit_ingredient::*;

#[derive(Debug, Clone, PartialEq, Eq)]
#[derive(Serialize, Deserialize)]
#[serde(untagged)]
pub enum TechnologyMaxLevel {
    Finite(u64),
    Infinite(TechnologyInfiniteLiteral),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[derive(Serialize, Deserialize)]
pub enum TechnologyInfiniteLiteral {
    #[serde(rename = "infinite")]
    Infinite,
}

/// 一个科技原型。
#[derive(Debug, Clone, PartialEq, Eq)]
#[derive(Serialize, Deserialize)]
pub struct Technology {
    /// 科技名称。
    ///
    /// 这是科技表的自然主键，也是 `prerequisites` / `successors` 自引用关系的目标键。
    pub name: String,

    /// 所属展示大分组。
    pub group: String,

    /// 所属展示子分组。
    pub subgroup: String,

    /// 展示排序键。
    pub order: String,

    /// 是否隐藏。
    pub hidden: bool,

    /// 是否为关键科技。
    pub essential: bool,

    /// 是否默认启用。
    pub enabled: bool,

    /// 禁用时是否仍在 UI 中可见。
    pub visible_when_disabled: bool,

    /// 是否为升级链科技。
    pub upgrade: bool,

    /// 当前科技等级。
    pub level: u64,

    /// 最大科技等级。
    ///
    /// `prototype-api` 允许这里出现 `"infinite"`，因此 DTO 直接承接整数 / 字符串联合形态。
    pub max_level: TechnologyMaxLevel,

    /// 研究所需单位数。
    pub research_unit_count: u64,

    /// 每个研究单位耗时。
    pub research_unit_energy: u64,

    /// 研究所需科技包列表。
    ///
    /// `ResearchUnitIngredient.name` 会引用 `items.name`，是科技与物品之间的重要外键。
    #[serde(default)]
    pub research_unit_ingredients: ArrayOrEmptyObject<ResearchUnitIngredient>,

    /// 特殊研究触发器。
    ///
    /// 某些科技不是靠常规科研包推进，而是靠行为触发。
    pub research_trigger: Option<ResearchTrigger>,

    /// 前置科技列表。
    ///
    /// 这是面向同表 `technologies.name` 的自引用多对多关系来源。
    #[serde(default)]
    pub prerequisites: ArrayOrEmptyObject<String>,

    /// 科技效果列表。
    ///
    /// `unlock-recipe`、`unlock-space-location` 等效果会把科技与其他主表连接起来。
    #[serde(default)]
    pub effects: ArrayOrEmptyObject<TechnologyEffect>,

    /// 该科技对应的配方是否允许生产力。
    pub allows_productivity: bool,

    /// 后继科技列表。
    ///
    /// 这是面向同表 `technologies.name` 的反向自引用关系来源。
    #[serde(default)]
    pub successors: ArrayOrEmptyObject<String>,

    /// 无限科技或等级链科技的研究单位公式。
    pub research_unit_count_formula: Option<String>,
}
